flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/MiniDispatcher.java

```diff
@@ -124,16 +125,23 @@ protected CleanupJobState jobReachedTerminalState(ExecutionGraphInfo executionGr
                 executionGraphInfo.getArchivedExecutionGraph();
         final CleanupJobState cleanupHAState = super.jobReachedTerminalState(executionGraphInfo);
 
-        if (jobCancelled || executionMode == ClusterEntrypoint.ExecutionMode.DETACHED) {
+        JobStatus jobStatus = archivedExecutionGraph.getState();
+        if ((jobStatus != null && jobStatus.isGloballyTerminalState())
```

Comment:
Can the job status be null here? My intuition would be that for the terminal state this should never be the case and would actually signal an underlying issue. Maybe adding a safeguard (`Objects.requireNonNull`) with a reasonable message could a better fit?